OCD-4160: Questionable Activity Cleanup

kekey1 opened the pull request.

kekey1 committed:
feat!: Remove updateParticpantsJob; Add fixupQuestionableActivityJob

[#OCD-4160]

kekey1 committed:
feat: Record activity ID with each new questionable activity entry

Someday we might like to link back between the things we think were
questionable and the user actions, or activities, that generated those
things. Just setting up the link.

[#OCD-4160]

kekey1 committed:
Merge branch 'staging' into OCD-4160

kekey1 committed:
feat: Make fixup QA job associate activity ID to each QA

[#OCD-4160]

kekey1 committed:
Merge branch 'staging' into OCD-4160

kekey1 committed:
Merge branch 'staging' into OCD-4160

kekey1 committed:
feat: Adjust activity search timeframe for each questionable activity

Search for activity from the beginning to the end of the second that the
questionable actiivty was recorded (so, same second but from 0-999
milliseconds).

[#OCD-4160]

kekey1 committed:
Merge branch 'staging' into OCD-4160

kekey1 committed:
fix!: Adjust questionable activity for criteria added/removed

The activity should not be counted as questionable if the user is
attesting to a Cures criteria in the same action as unattesting to the
analogous original criteria.

[#OCD-4160]

kekey1 committed:
feat: Reprocess criteria added/removed questionable activities in job

The job soft-deletes existing quesitonable activities of this type and
then reprocesses activity to correctly add things that are questionable.
We are ignoring criteria added/removed activity when the activity added
a Cures criteria in favor of it's original-criteria counterpart.

[#OCD-4160]

kekey1 committed:
feat: Save 'reason' field whenever it is available for any activity

We were saving the 'reason' when recording questionable activity, but
a user could potentially be entering reasons for changing data at times
when the change isn't questionable. In that case we are losing whatever
they've put in. I want to save it with the activity itself.

[#OCD-4160]

kekey1 committed:
feat: Correctly identify non-active certificate edited questionable act

This was wrong in some cases and sometimes appears related to
"Certification Date Changed" activity being incorrect as well.

[#OCD-4160]

kekey1 committed:
feat: Rename Current Certification Date Changed trigger

[#OCD-4160]

kekey1 committed:
feat: Add checks for "certification date changed" questionable activity

[#OCD-4160]

kekey1 committed:
Merge branch 'staging' into OCD-4160

kekey1 committed:
Merge branch 'staging' into OCD-4160

kekey1 committed:
feat: Reprocess cert status history questionable activities

[#OCD-4160]

andlar commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/domain/CertificationStatusEvent.java:
```diff
@@ -93,9 +93,17 @@ public void setStatus(final CertificationStatus status) {
+        return reason;
+    }
+
+    public void setReason(final String reason) {
```
I _think_ we were getting rid of the `final` descriptors, but am not positive

andlar commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/questionableactivity/listing/UpdatedCertificationStatusDate.java:
Not sure what's going on here. Is this a type of QA that's going away? It seems like there is one based on the comment in the QA concept file. Do we still need this file?

kekey1 replied:
This one was named in a really misleading way. It is not that the Certification Date was updated (I added a new type of questionable activity for that). It is that the date of the most recent certification status event was updated. This might have been the date of being Withdrawn being updated and not necessarily the date of being Active (the Certification Date). I changed the name of this activity to be more clear elsewhere in the feature. And all the cases that this was checking for are already handled by other triggers (I think - we'll see if we find anything missing during testing!)

andlar commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/questionableActivity/FixupQuestionableActivity.java:
```diff
@@ -0,0 +1,225 @@
+        //new Date() as the Questionable Activity date so it would be milliseconds off of the Activity date.
+        //I think querying for activity events during the same second is acceptable especially since
+        //we are not proceeding if there is more than 1 match.
+        //The results are MUCH better when doing it this way.
```
🤣🤣🤣

kekey1 committed:
test: Choose times for unit tests that aren't 4-5 hours before next day

[#OCD-4160]

kekey1 committed:
style: Remove 'final' modifiers in getters/setters

[#OCD-4160]

kekey1 committed:
refactor: Move no longer used QA checker to onetime job package

This move allows us to keep the job code the same but will help us
remember to clean up the unused QA checker in the future.

[#OCD-4160]

kekey1 committed:
Merge branch 'staging' into OCD-4160

kekey1 committed:
Merge branch 'staging' into OCD-4160

kekey1 committed:
fix!: Make copy of cert status events to check questionable activity

[#OCD-4160]

kekey1 committed:
fix: Preserve cert status change reason on certification date change

[#OCD-4160]
